projects/timer_challenge/src/main.c

```diff
@@ -0,0 +1,47 @@
+#include <stdint.h>      // for integer types
+#include "interrupt.h"   // interrupts are required for soft timers
+#include "log.h"         // for printing
+#include "soft_timer.h"  // for soft timers
+#include "wait.h"        // for wait function
+
+#define A_INC_PERIOD_MS 500  // milliseconds between a-increment
+
+typedef struct Counters {
+  uint16_t counter_a;
+  uint16_t counter_b;
+} Counters;
+
+void prv_timer_callback(SoftTimerId timer_id, void *context) {
+  Counters *storage = context;
+
+  // increment a
+  ++storage->counter_a;
+  LOG_DEBUG("Counter A: %i\n", storage->counter_a);
+
+  // increment b every other time a is incremented
+  if (storage->counter_a % 2 == 0) {
+    ++storage->counter_b;
+    LOG_DEBUG("Counter B: %i\n", storage->counter_b);
+  }
+
+  // start the timer again
+  soft_timer_start_millis(A_INC_PERIOD_MS, prv_timer_callback, storage, NULL);
+}
+
+int main() {
+  interrupt_init();   // interrupts must be initialized for soft timers to work
+  soft_timer_init();  // soft timers must be initialized before using them
+
+  Counters storage = { 0 };  // we use this to initialize a struct to be all 0
+
+  soft_timer_start_millis(A_INC_PERIOD_MS,     // timer duration
+                          prv_timer_callback,  // function to call after timer
+                          &storage,            // automatically gets cast to void*
+                          NULL);               // timer id - not needed here
+
+  while (true) {
+    wait();  // waits until an interrupt is triggered rather than endlessly spinning
```

Comment:
I really like how you have comments in many of these lines.
However, for actual projects, we generally don't add these comments unless it is needed to clarify things.
A lot of it is self-explanatory

eg I think the below is sort of obvious
```
  // start the timer again
  soft_timer_start_millis(A_INC_PERIOD_MS, prv_timer_callback, storage, NULL);
```

Again, I like how you've added a lot of comments for your learning. However, when it comes to actual projects, remeber that the more comments you add, the more the code must be maintained in case there are changes.
It can also clutter the code if there are too many comments that are self-explanatory.